File: swift-bridge/Sources/SpriteKitBridge/PhysicsWorld.swift
import SpriteKit

@_cdecl("sk_physics_world_get_gravity_dx")
public func sk_physics_world_get_gravity_dx(_ worldHandle: UnsafeMutableRawPointer?) -> Double {
    guard let world: SKPhysicsWorld = skBorrow(worldHandle) else { return 0 }
    return Double(world.gravity.dx)
}

@_cdecl("sk_physics_world_get_gravity_dy")
public func sk_physics_world_get_gravity_dy(_ worldHandle: UnsafeMutableRawPointer?) -> Double {
    guard let world: SKPhysicsWorld = skBorrow(worldHandle) else { return -9.8 }
    return Double(world.gravity.dy)
}

@_cdecl("sk_physics_world_set_gravity")
public func sk_physics_world_set_gravity(_ worldHandle: UnsafeMutableRawPointer?, _ dx: Double, _ dy: Double) {
    guard let world: SKPhysicsWorld = skBorrow(worldHandle) else { return }
    world.gravity = CGVector(dx: dx, dy: dy)
}

@_cdecl("sk_physics_world_get_speed")
public func sk_physics_world_get_speed(_ worldHandle: UnsafeMutableRawPointer?) -> Double {
    guard let world: SKPhysicsWorld = skBorrow(worldHandle) else { return 1 }
    return Double(world.speed)
}

@_cdecl("sk_physics_world_set_speed")
public func sk_physics_world_set_speed(_ worldHandle: UnsafeMutableRawPointer?, _ speed: Double) {
    guard let world: SKPhysicsWorld = skBorrow(worldHandle) else { return }
    world.speed = CGFloat(speed)
}

@_cdecl("sk_physics_world_body_at_point")
public func sk_physics_world_body_at_point(_ worldHandle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double) -> UnsafeMutableRawPointer? {
    guard let world: SKPhysicsWorld = skBorrow(worldHandle), let body = world.body(at: CGPoint(x: x, y: y)) else { return nil }
    return skRetain(body)
}

@_cdecl("sk_physics_world_body_in_rect")
public func sk_physics_world_body_in_rect(_ worldHandle: UnsafeMutableRawPointer?, _ x: Double, _ y: Double, _ width: Double, _ height: Double) -> UnsafeMutableRawPointer? {
    guard let world: SKPhysicsWorld = skBorrow(worldHandle), let body = world.body(in: CGRect(x: x, y: y, width: width, height: height)) else { return nil }
    return skRetain(body)
}

@_cdecl("sk_physics_world_body_along_ray")
public func sk_physics_world_body_along_ray(_ worldHandle: UnsafeMutableRawPointer?, _ startX: Double, _ startY: Double, _ endX: Double, _ endY: Double) -> UnsafeMutableRawPointer? {
    guard let world: SKPhysicsWorld = skBorrow(worldHandle), let body = world.body(alongRayStart: CGPoint(x: startX, y: startY), end: CGPoint(x: endX, y: endY)) else { return nil }
    return skRetain(body)
}

File: src/physics_world.rs
use apple_cf::cg::{CGPoint, CGRect, CGVector};

use crate::ffi;
use crate::physics_body::PhysicsBody;
use crate::physics_contact::PhysicsContactDelegate;
use crate::private::handle_type;

handle_type!(PhysicsWorld);

impl PhysicsWorld {
    #[must_use]
    pub fn gravity(&self) -> CGVector {
        CGVector::new(
            unsafe { ffi::sk_physics_world_get_gravity_dx(self.ptr) },
            unsafe { ffi::sk_physics_world_get_gravity_dy(self.ptr) },
        )
    }

    pub fn set_gravity(&self, gravity: CGVector) {
        unsafe { ffi::sk_physics_world_set_gravity(self.ptr, gravity.dx, gravity.dy) };
    }

    #[must_use]
    pub fn speed(&self) -> f64 {
        unsafe { ffi::sk_physics_world_get_speed(self.ptr) }
    }

    pub fn set_speed(&self, speed: f64) {
        unsafe { ffi::sk_physics_world_set_speed(self.ptr, speed) };
    }

    #[must_use]
    pub fn body_at_point(&self, point: CGPoint) -> Option<PhysicsBody> {
        unsafe {
            PhysicsBody::from_raw(ffi::sk_physics_world_body_at_point(
                self.ptr, point.x, point.y,
            ))
        }
    }

    #[must_use]
    pub fn body_in_rect(&self, rect: CGRect) -> Option<PhysicsBody> {
        unsafe {
            PhysicsBody::from_raw(ffi::sk_physics_world_body_in_rect(
                self.ptr,
                rect.x,
                rect.y,
                rect.width,
                rect.height,
            ))
        }
    }

    #[must_use]
    pub fn body_along_ray(&self, start: CGPoint, end: CGPoint) -> Option<PhysicsBody> {
        unsafe {
            PhysicsBody::from_raw(ffi::sk_physics_world_body_along_ray(
                self.ptr, start.x, start.y, end.x, end.y,
            ))
        }
    }

    pub fn set_contact_delegate(&self, delegate: Option<&PhysicsContactDelegate>) {
        unsafe {
            ffi::sk_physics_world_set_contact_delegate(
                self.ptr,
                delegate.map_or(core::ptr::null_mut(), PhysicsContactDelegate::as_ptr),
            );
        };
    }

    #[must_use]
    pub fn has_contact_delegate(&self) -> bool {
        unsafe { ffi::sk_physics_world_has_contact_delegate(self.ptr) }
    }
}
